app/controllers/application_controller.rb

```diff
@@ -57,18 +57,25 @@ def user_fully_authenticated?
     user_signed_in? && current_user.two_factor_enabled? && is_fully_authenticated?
   end
 
-  def confirm_two_factor_setup
+  def confirm_two_factor_authenticated
+    authenticate_user!
+
     user_decorator = UserDecorator.new(current_user)
-    return if user_decorator.may_bypass_two_factor_setup?(session)
 
+    return if user_decorator.may_bypass_2fa?(session) || user_fully_authenticated?
+
+    return prompt_to_set_up_2fa unless current_user.two_factor_enabled?
+
+    prompt_to_enter_otp
```

Comment:
Before I discovered the Ruby Style Guide, I was using if-else statements like you, but then learned about the guard clause, and now I like it better and can read it just as easily, if not better than the nested conditionals. Maybe it's just a matter of time for you.

The point of the guard clause is to bail out of the method as early as possible, when you know there's no point in checking more conditionals. I also like that it's only 5 lines of code versus 10.

Here is the section about guard clauses in the style guide: https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals